### PR TITLE
Add vector mass and vector diffusion PA integrators

### PR DESCRIPTION
**Pull request by jandrej**
This adds PA for `VectorMassIntegrator` and `VectorDiffusionIntegrator`.

~Beware, there are no unit tests for this right now.~

- [x] Code builds.
- [x] Code passes make style.
- [x] Update CHANGELOG.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/1167 | @jandrej | @v-dobrev | @barker29 + @jakubcerveny | 12/04/19 | 12/30/19 | ⌛due 01/06/20 |

### CODE REVIEW

File: tests/unit/fem/test_pa_kernels.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+   blf_pa.Mult(x, y_pa);
+
+   y_fa -= y_pa;
+   double difference = y_fa.Norml2();
+
+   delete mesh;
+   return difference;
+}
+
+TEST_CASE("Nonlinear Convection", "[PartialAssembly], [NonlinearPA]")
```
Comment (jandrej) [inline tests/unit/fem/test_pa_kernels.cpp]: Copy-pasta error detected.

**Comment by tzanio** [comment]
@jandrej, can you take a look at the Appveyor errors?

**Comment by jandrej** [comment]
I'm on the case.

File: tests/unit/fem/test_pa_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#include <fstream>
+#include <iostream>
+
+using namespace mfem;
+
+namespace pa_kernels
+{
+
+template <typename INTEGRATOR>
+double test_vector_pa_integrator(int dim)
```
Comment (barker29) [inline tests/unit/fem/test_pa_kernels.cpp]: This is a nice test.

File: fem/bilininteg_vecdiffusion.cpp [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+static void PAVectorDiffusionSetup(const int dim,
+                                   const int D1D,
+                                   const int Q1D,
+                                   const int NE,
+                                   const Array<double> &W,
+                                   const Vector &J,
+                                   const double COEFF,
+                                   Vector &op)
+{
+   if (dim == 1) { MFEM_ABORT("dim==1 not supported in PADiffusionSetup"); }
```
Comment (barker29) [inline fem/bilininteg_vecdiffusion.cpp]: Catch crazy dims (0, 4 etc.)

File: fem/bilininteg.hpp [hunk 4]
```diff
@@ -1835,6 +1835,11 @@ class VectorMassIntegrator: public BilinearFormIntegrator
    Coefficient *Q;
    VectorCoefficient *VQ;
    MatrixCoefficient *MQ;
+   // PA extension
```
Comment (barker29) [inline fem/bilininteg.hpp]: May be out of scope for this PR, but I am wondering if the next four lines (which are copied verbatim in at least four places) should be in the base class? (Or maybe in a separate PA base class, with multiple inheritance, which is probably a mistake.) I just don't like seeing these member data rewritten everytime we implement a PA `BilinearForm`.

**Comment by jandrej** [comment]
I addressed Andrews comments.

File: tests/unit/fem/test_pa_kernels.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+   blf_fa.Finalize();
+   blf_fa.Mult(x, y_fa);
+
+   BilinearForm blf_pa(&fes);
+   blf_pa.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+   blf_pa.AddDomainIntegrator(new INTEGRATOR);
+   blf_pa.Assemble();
+   blf_pa.Mult(x, y_pa);
+
+   y_fa -= y_pa;
```
Comment (jakubcerveny) [inline tests/unit/fem/test_pa_kernels.cpp]: I like this too!

File: fem/bilininteg_vecmass.cpp [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+                              const int NE,
+                              const Array<double> &B,
+                              const Array<double> &Bt,
+                              const Vector &op,
+                              const Vector &x,
+                              Vector &y)
+{
+   if (dim == 2)
+   {
+      return PAVectorMassApply2D(NE, B, Bt, op, x, y, D1D, Q1D);
```
Comment (jakubcerveny) [inline fem/bilininteg_vecmass.cpp]: Is this enough? Standard diffusion has optimized cases for the most common orders here. (Just wondering -- I'm not sure what the exact goal of the PR is.)

**Comment by tzanio** [comment]
Merged in `next` for testing ...

**Comment by tzanio** [comment]
@jandrej, please update `CHANGELOG`

File: fem/bilininteg_vecmass.cpp [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+   dofs1D = maps->ndof;
+   quad1D = maps->nqpt;
+   pa_data.SetSize(ne*nq, Device::GetMemoryType());
+   double coeff = 1.0;
+   if (Q)
+   {
+      ConstantCoefficient *cQ = dynamic_cast<ConstantCoefficient*>(Q);
+      MFEM_VERIFY(cQ != NULL, "Only ConstantCoefficient is supported.");
+      coeff = cQ->constant;
+   }
```
Comment (artv3) [inline fem/bilininteg_vecmass.cpp]: Could we also add support for non-constant coefficients here? as was done in ``bilininteg_mass.cpp``

Comment (jandrej) [reply fem/bilininteg_vecmass.cpp]: I'd like to postpone this and do rigorous testing on a few implementations first.

File: fem/bilininteg_vecmass.cpp (hunk 6 above)
Comment (jandrej) [reply fem/bilininteg_vecmass.cpp]: The reason for this is the increased compile time. The users who really need the performance increase will have to define the template parameters (for now).

Comment (jakubcerveny) [reply fem/bilininteg_vecmass.cpp]: Should we prepare the switch with the templated cases and just #ifdef it out, so the user can just turn the optimized cases on?

Comment (jandrej) [reply fem/bilininteg_vecmass.cpp]: If we would do that, we have to do it on every other kernel as well to be consistent. It is a good idea and we should consider that.

Comment (jakubcerveny) [reply fem/bilininteg_vecmass.cpp]: Better yet, the #ifdef could be controlled by the configuration. For development (e.g., "make debug"), you don't want long compile times but for deployment, long compile should be worth the performance improvement...

Comment (jandrej) [reply fem/bilininteg_vecmass.cpp]: Are you ok to open an issue for that or do you want me to do it? I will bring it up in the next meeting as well.

Comment (jakubcerveny) [reply fem/bilininteg_vecmass.cpp]: It's just an idea, let's see what the editors think.

Comment (tzanio) [reply fem/bilininteg_vecmass.cpp]: We decided that we need a separate discussion about how we generally address this. We will keep that going in a new issue, and will merge this PR as is.

**Pull request merged**